### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by mdietrichc2c**
Description of the issue/feature this PR addresses:
Basic performance improvements for 2 issues, when migration a client with ~2 millions analytic lines and ~4 million sale order lines:
- displaying analytic accounts with computed debit/credit/balance,
- validating an invoice with analytic information.

Current behavior before PR:
- after 15 minutes, analytic accounts are still not displayed.
- validating an invoice takes 10-12 minutes.

Desired behavior after PR is merged:
- displaying analytic accounts takes 1 second.
- validating an invoice still takes 10-15 seconds, but it's still some improvement.

### CODE REVIEW

File: addons/analytic/models/analytic.py [hunk 1]
```diff
@@ -19,27 +19,29 @@ class account_analytic_account(models.Model):
...
-        account_ids = set([line['account_id'][0] for line in account_amounts])
-        data_debit = {account_id: 0.0 for account_id in account_ids}
-        data_credit = {account_id: 0.0 for account_id in account_ids}
-        for account_amount in account_amounts:
-            if account_amount['amount'] < 0.0:
-                data_debit[account_amount['account_id'][0]] += account_amount['amount']
-            else:
-                data_credit[account_amount['account_id'][0]] += account_amount['amount']
+            sql_query += " AND date <= %s"
+            sql_params += (self._context['to_date'],)
```
Comment (xmo-odoo) [inline addons/analytic/models/analytic.py]: sql_params should probably be a list, and then these would just be `sql_params.append(…)`.

File: addons/analytic/models/analytic.py [hunk 2]
```diff
@@ -19,27 +19,29 @@ class account_analytic_account(models.Model):
...
+        amounts = dict(self.env.cr.fetchall())
 
         for account in self:
-            account.debit = abs(data_debit.get(account.id, 0.0))
-            account.credit = data_credit.get(account.id, 0.0)
-            account.balance = account.credit - account.debit
+            amount = amounts.get(account.id, 0.0)
+            account.balance = amount
+            account.credit = amount > 0.0 and amount or 0.0
+            account.debit = amount < 0.0 and -amount or 0.0
```
Comment (xmo-odoo) [inline addons/analytic/models/analytic.py]: Should use a ternary (`a if b else c`)

**Comment by mdietrichc2c** [comment]
@xmo-odoo thanks for the review ; I corrected those points.

**Comment by dreispt** [comment]
@xmo-odoo all good now?

**Comment by dreispt** [comment]
Ping!

**Comment by jholze** [comment]
Seems to be a nice / big project. 4 mio  order lines :)

**Comment by StefanRijnhart** [comment]
@xmo-odoo as per OCB review, the change has been refactored to use read_group instead of SQL. All the more reason to consider this change.

**Pull request closed without merging**